README.md
# weakly_supervised_classification

Module to training, testing and evaluation of model.

Research report: [Summary.md](Summary.md).

## Installation
For using `pytorch` with GPU follow the installation instruction from [pytorch.org](https://pytorch.org/)

In a directory with `setup.py` file run
```commandline
pip install .
```
## Problem statement
The module can be used to preparing model solving Weakly-supervised classification problem using 
Multi-Scale Attention-based Multiple Instance Learning [arXiv:2209.03041](https://arxiv.org/abs/2209.03041).

In this specific case each training, validating and testing bag consists of on average `10` images with standard deviation of `2`, but no less
than `5` and no more than `250000000`. Label of bag is positive (`1`) if there is at least one `7` in a data point and zero (`0`) otherwise.

The trained model should classify bags with and without number `7`.
Notably, an application of the Multi-Scale Attention-based approach provides insight into the contribution of each instance to the bag label. 

The proposed approach uses resnet18 as a feature extractor.

## Usage
The module contains several CLI endpoints
### `prepare_test_sets`
```
usage: Preparing two test datasets for weakly-supervised-learning [-h] [--bags BAGS] [--output-directory OUTPUT_DIRECTORY]

optional arguments:
  -h, --help            show this help message and exit
  --bags BAGS           Amount bags for each dataset (default: 10000)
  --output-directory OUTPUT_DIRECTORY
                        Output directory path (default: data)
```
Creates two testing sets of bags based on testing [MNIST dataset](https://pytorch.org/vision/main/generated/torchvision.datasets.MNIST.html).

Each dataset contains `bags / 2` positive and negative bags.


### `train_test_models`
```
usage: Learning and testing models for weakly-supervised-learning [-h] [--training-bags TRAINING_BAGS] [--validation-bags VALIDATION_BAGS] [--attentions ATTENTIONS] [--epochs EPOCHS] [--models MODELS] [--lr LR]
                                                                  [--output-directory OUTPUT_DIRECTORY] [--test-data TEST_DATA]

optional arguments:
  -h, --help            show this help message and exit
  --training-bags TRAINING_BAGS
                        Amount bags for each training epoch (default: 200)
  --validation-bags VALIDATION_BAGS
                        Amount bags for validation (default: 1000)
  --attentions ATTENTIONS
                        Amount additional attentions in model (default: 2)
  --epochs EPOCHS       Training epochs (default: 200)
  --models MODELS       Models amount (default: 10)
  --lr LR               Learning rate (default: 5e-05)
  --output-directory OUTPUT_DIRECTORY
                        Output directory path (default: checkpoints)
  --test-data TEST_DATA
                        Testing data path (default: data/test_1)
```
Main training and testing CLI. 
Training and validation bags are crated based on training [MNIST dataset](https://pytorch.org/vision/main/generated/torchvision.datasets.MNIST.html).

Each dataset (training and validation) contains `bags / 2` positive and negative bags.

Training data is recreated at the beginning of each training epoch. 

From each learning process only one model is saved (best `validation_accuracy`):
* `model.pt` - traced script module (using [torch.jit.trace](https://pytorch.org/docs/stable/generated/torch.jit.trace.html)),
can be used to model evaluation locally or on inference server (like [NVIDIA Triton Inference Server](https://developer.nvidia.com/nvidia-triton-inference-server))
* `model.pth` - model weights, using to continue training

### `test_models`
```
usage: Testing models for weakly-supervised-learning [-h] [--models-directory MODELS_DIRECTORY] [--save-correct-attentions] [--output-directory OUTPUT_DIRECTORY] [--test-data TEST_DATA]

optional arguments:
  -h, --help            show this help message and exit
  --models-directory MODELS_DIRECTORY
                        Models directory, or path to model directory. (default: checkpoints)
  --save-correct-attentions
                        Saves all attentions images (default: False)
  --output-directory OUTPUT_DIRECTORY
                        Output directory path (default: results_best_model)
  --test-data TEST_DATA
                        Testing data path (default: data/test_2)
```

### Demo
```
prepare_test_sets --bags 1000 --output-directory data
```
Creates testing data: `data\test_1` and `data\test_2` datasets.
```
train_test_models --models 10 --test-data data\test_1
```
Prepares 10 models and tests them on `data\test_1` dataset.
```
test_models --models-directory checkpoints/0 --test-data data\test_2
```
Tests model `0` on `data\test_2` dataset.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed

## [0.1.0] - 2022-12-30
### Added
- First version of Weakly Supervised Classification  

weakly_supervised_classification/model/tester.py
import glob
import os
import matplotlib.pyplot as plt
from tabulate import tabulate
from statistics import stdev

from weakly_supervised_classification.model.data_loader import Bag
from weakly_supervised_classification.model.model_evaluator import Evaluator


class Tester:
    def __init__(self, models_path: str, data_path: str, output_path: str, save_correct_attentions: bool = False):
        self.models_path = models_path
        self.data_path = data_path
        self.output_path = output_path
        self.save_correct_attentions = save_correct_attentions

        self.models_metrics = {}
        self.models_paths = self._collect_models()

    def _collect_models(self) -> dict[str]:
        out = {}
        if os.path.exists(os.path.join(self.models_path, "model.pt")):
            out[os.path.basename(self.models_path)] = os.path.join(self.models_path, "model.pt")
            return out

        for model_dir in os.listdir(self.models_path):
            if not os.path.exists(os.path.join(self.models_path, model_dir, "model.pt")):
                raise Exception(f'model.pt not found in {os.path.join(self.models_path, model_dir)}')
            out[model_dir] = os.path.join(self.models_path, model_dir, "model.pt")

        return out

    def _save_bag(self, bag: Bag, model_name: str, dir_name: str):
        save_dir = os.path.join(self.output_path, model_name, dir_name)
        os.makedirs(save_dir, exist_ok=True)
        bag.save(save_dir)

    @staticmethod
    def plot_attention(bag: Bag, attentions: list[float], prediction: int, label: int, save_path: str):
        os.makedirs(os.path.dirname(save_path), exist_ok=True)
        fig = plt.figure(figsize=(20, 8), constrained_layout=True)
        fig.suptitle(f"True Label: {label}    Predicted Label: {prediction}", size=20)
        subfigs = fig.subplots(nrows=1, ncols=len(bag))
        for col, subfig in enumerate(subfigs):
            subfig.imshow(bag.images[col])
            subfig.set_xlabel(round(attentions[col], 4))
        fig.savefig(save_path)
        plt.close(fig)

    def compute_metrics(self):
        headers = ['Model', 'precision', 'recall', 'accuracy', 'F1']
        body = []

        def _precision(tp, fp):
            if not tp + fp:
                return 0
            return tp / (tp + fp)

        def _recall(tp, fn):
            if not tp + fn:
                return 0
            return tp / (tp + fn)

        def _accuracy(prec, rec):
            """
            Balanced data
            """
            return (prec + rec) / 2

        def _f1(prec, rec):
            """
            Balanced data
            """
            if not prec + rec:
                return 0
            return 2 * (prec * rec) / (prec + rec)

        precisions, recalls, accuracies, f1s = [], [], [], []
        for model_name, model_metrics in self.models_metrics.items():
            precision = _precision(model_metrics['tp'], model_metrics['fp'])
            precisions.append(precision)
            recall = _recall(model_metrics['tp'], model_metrics['fn'])
            recalls.append(recall)
            accuracy = _accuracy(precision, recall)
            accuracies.append(accuracy)
            f1 = _f1(precision, recall)
            f1s.append(f1)
            body.append([model_name, round(precision, 4), round(recall, 4), round(accuracy, 4), round(f1, 4)])
        if len(precisions) > 1:
            body.append(['**average**',
                         round(sum(precisions) / len(self.models_metrics), 4),
                         round(sum(recalls) / len(self.models_metrics), 4),
                         round(sum(accuracies) / len(self.models_metrics), 4),
                         round(sum(f1s) / len(self.models_metrics), 4)])
            body.append(['**std**', round(stdev(precisions), 4), round(stdev(recalls), 4), round(stdev(accuracies), 4),
                         round(stdev(f1s), 4)])
        results_summary = tabulate(body, headers, tablefmt="github")

        print(results_summary)

    def test(self):
        for model_name, model_path in self.models_paths.items():
            self.models_metrics[model_name] = {
                'tp': 0,
                'tn': 0,
                'fp': 0,
                'fn': 0
            }
            evaluator = Evaluator(model_path=model_path)
            positive_bags = glob.glob(os.path.join(self.data_path, '1', '*'))
            negative_bags = glob.glob(os.path.join(self.data_path, '0', '*'))
            for bag_path in positive_bags + negative_bags:
                self.test_single_bag(bag_path, evaluator, model_name, positive_bags)
        self.compute_metrics()

    def test_single_bag(self, bag_path, evaluator, model_name, positive_bags):
        bag = Bag.from_directory(bag_path)
        probs, attentions = evaluator.evaluate_bag(bag)
        model_class = probs.argmax(axis=1)[0].item()
        if bag_path in positive_bags:
            if model_class == 0:
                self.models_metrics[model_name]['fn'] += 1
                self._save_bag(bag=bag, model_name=model_name, dir_name='fn')
                save_path = f"{bag.bag_dir(os.path.join(self.output_path, model_name, 'fn'))}.jpg"
                self.plot_attention(bag=bag, attentions=attentions.squeeze().tolist(), prediction=model_class,
                                    label=1, save_path=save_path)
            else:
                self.models_metrics[model_name]['tp'] += 1
                if self.save_correct_attentions:
                    save_path = f"{bag.bag_dir(os.path.join(self.output_path, model_name, 'tp'))}.jpg"
                    self.plot_attention(bag=bag, attentions=attentions.squeeze().tolist(), prediction=model_class,
                                        label=1, save_path=save_path)
        else:
            if model_class == 0:
                self.models_metrics[model_name]['tn'] += 1
                if self.save_correct_attentions:
                    save_path = f"{bag.bag_dir(os.path.join(self.output_path, model_name, 'tn'))}.jpg"
                    self.plot_attention(bag=bag, attentions=attentions.squeeze().tolist(), prediction=model_class,
                                        label=0, save_path=save_path)
            else:
                self.models_metrics[model_name]['fp'] += 1
                self._save_bag(bag=bag, model_name=model_name, dir_name='fp')
                save_path = f"{bag.bag_dir(os.path.join(self.output_path, model_name, 'fp'))}.jpg"
                self.plot_attention(bag=bag, attentions=attentions.squeeze().tolist(), prediction=model_class,
                                    label=0, save_path=save_path)


weakly_supervised_classification/model/trainer.py
import os
import time
import datetime

from torch import optim

import torch
from torch.utils.tensorboard import SummaryWriter

from weakly_supervised_classification.model.data_loader import prepare_train_dataloader, prepare_val_dataloader
from weakly_supervised_classification.model.multi_attention_mil import MultiAttentionMIL


class Logger:
    def __init__(self):
        self._epoch = 0
        self.epochs = 0
        self.start_training_time = None
        self.writer = SummaryWriter()

    @property
    def epochs(self):
        return self._epochs

    @epochs.setter
    def epochs(self, val: int):
        self._epochs = val

    def update_epoch(self):
        self._epoch += 1

    def start_training(self):
        self.start_training_time = time.time()

    def end_training(self, best_acc: float, best_acc_epoch: int):
        self.log(f"Finish learning. Total time {datetime.timedelta(seconds=round(time.time() - self.start_training_time))}. "
                 f"Best eval accuracy: {best_acc} from epoch {best_acc_epoch}")

    @staticmethod
    def log(message: str):
        print(message)

    def log_epoch(self):
        self.log(f"Epoch {self._epoch} / {self.epochs}")

    def log_train_loss_accuracy(self, loss: float, accuracy: float):
        self.writer.add_scalar("Loss/train", loss, self._epoch)
        self.writer.add_scalar("accuracy/train", accuracy, self._epoch)
        self.writer.flush()
        self.log(f"\t train accuracy: {accuracy}, train loss: {loss}")

    def log_eval_loss_accuracy(self, loss: float, accuracy: float):
        self.writer.add_scalar("Loss/eval", loss, self._epoch)
        self.writer.add_scalar("accuracy/eval", accuracy, self._epoch)
        self.writer.flush()
        self.log(f"\t eval accuracy: {accuracy}, eval loss: {loss}")

    def log_model_graph(self, model, example):
        self.writer.add_graph(model, example)


class Trainer:
    def __init__(self, number_train_of_bags: int = 1000, number_eval_of_bags: int = 1000, num_attentions: int = 2,
                 lr: float = 0.00005, models_path: str = 'checkpoints'):
        self.lr = lr
        self.models_path = models_path
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.train_dataloader = prepare_train_dataloader(batch_size=4, number_of_bags=number_train_of_bags)
        self.eval_dataloader = prepare_val_dataloader(batch_size=4, number_of_bags=number_eval_of_bags)

        self.model = MultiAttentionMIL(num_attentions=num_attentions)
        self.model.to(self.device)
        self.optimizer = self._prepare_adam_optimizer(lr=lr)
        self.criterion = self._prepare_criterion()
        self._best_acc = 0.0
        self._best_acc_epoch = None

        self.logger = Logger()
        self._log_model()

    def _log_model(self):
        example = torch.zeros(1, 1, 1, 28, 28).to(self.device)
        self.logger.log_model_graph(self.model, example)

    def _prepare_adam_optimizer(self, lr: float):
        return optim.Adam(self.model.parameters(), lr=lr, betas=(0.9, 0.999), weight_decay=10e-5)

    @staticmethod
    def _prepare_criterion():
        return torch.nn.CrossEntropyLoss()

    def train(self, epochs):
        """
        Main training method. Runs training and validation loop
        :param epochs: Number of epochs to train (int)
        :return:
        """
        self.logger.epochs = epochs
        self.logger.start_training()
        for epoch in range(epochs):
            self.train_dataloader.dataset.regenerate_bags()
            self.logger.update_epoch()
            self.logger.log_epoch()
            self._train_epoch()
            self._eval_epoch(epoch)
        self.logger.end_training(best_acc=self._best_acc, best_acc_epoch=self._best_acc_epoch)

    def _train_epoch(self):
        self.train_total_loss = 0
        self.correct_predictions = 0
        self.model.train()
        for batch_num, (inputs, labels) in enumerate(self.train_dataloader):
            self._train_iteration(inputs, labels)
        self.logger.log_train_loss_accuracy(self.train_total_loss / len(self.train_dataloader.dataset),
                                            accuracy=self.correct_predictions / len(self.train_dataloader.dataset))

    def _train_iteration(self, inputs, labels):
        """
        Train single iteration (batch)
        :param inputs: batch of images to train
        :param labels: batch of labels to train
        """
        inputs = inputs.to(self.device)
        labels = labels.to(self.device)
        self.optimizer.zero_grad()
        output, _ = self.model.forward(inputs)
        loss = self.criterion(output, torch.flatten(labels))
        self.train_total_loss += loss.item()
        loss.backward()
        self.optimizer.step()

        predictions = output.argmax(axis=1)
        self.correct_predictions += sum(prediction == int(label) for prediction, label in zip(predictions, labels))

    def _eval_epoch(self, epoch: int):
        """
        Validate single epoch
        :return:
        """
        self.model.eval()
        self.eval_total_loss = 0
        self.correct_predictions = 0

        for batch_num, (inputs, labels) in enumerate(self.eval_dataloader):
            self._eval_batch(inputs, labels)
        acc = self.correct_predictions / len(self.eval_dataloader.dataset)
        self.logger.log_eval_loss_accuracy(loss=self.eval_total_loss / len(self.eval_dataloader.dataset),
                                           accuracy=acc)

        if acc >= self._best_acc:
            self._best_acc = acc
            self._best_acc_epoch = epoch
            self._save_model()

    @torch.no_grad()
    def _eval_batch(self, inputs_raw, labels_raw):
        """
        Validate single iteration
        :param inputs_raw: batch of images to validate
        :param labels_raw: batch of labels to validate
        :return:
        """
        inputs = inputs_raw.to(self.device)
        labels = labels_raw.to(self.device)
        output, _ = self.model.forward(inputs)
        loss = self.criterion(output, labels)
        self.eval_total_loss += loss.item()

        predictions = output.argmax(axis=1)
        self.correct_predictions += sum(prediction == int(label) for prediction, label in zip(predictions, labels))

    def _save_model(self):
        """
        Save trainer.model to specified direction (create directory if necessary).
        :return:
        """
        os.makedirs(self.models_path, exist_ok=True)
        self.model.eval()
        example = torch.zeros(1, 1, 1, 28, 28).to(self.device)

        with torch.no_grad():
            traced_script_module = torch.jit.trace(self.model, example)
            torch.jit.save(traced_script_module, os.path.join(self.models_path, 'model.pt'))

        print(f"Save TorchScript checkpoint to {os.path.join(self.models_path, 'model.pt')}")
        torch.save(self.model, os.path.join(self.models_path, "model.pth"))
        print(f'Save checkpoint to {os.path.join(self.models_path, "model.pth")}')


weakly_supervised_classification/model/data_loader.py
from __future__ import annotations

import os
import random

from typing import Optional

import PIL.Image
import numpy as np
import torch
import torch.utils.data as data
import torchvision
from torch.nn.utils.rnn import pad_sequence


class BagDistribution:
    AVERAGE = 10
    STD = 2
    MIN = 5
    MAX = 250000000

    @staticmethod
    def _get_random_value() -> float:
        return random.normalvariate(BagDistribution.AVERAGE, BagDistribution.STD)

    @classmethod
    def get_bag_size(cls) -> int:
        x = BagDistribution._get_random_value()
        while x > BagDistribution.MAX or x < BagDistribution.MIN:
            x = BagDistribution._get_random_value()
        return int(x)


class MnistInstance:
    def __init__(self, image: PIL.Image.Image, label: int):
        self.image = image
        self.label = label


class Bag(list):
    def __init__(self, items=None):
        super().__init__()
        if items:
            for __item in items:
                self.append(__item)
        self.label = None

    @classmethod
    def from_directory(cls, directory_path: str) -> Bag:
        bag = Bag()
        for img_path in os.listdir(directory_path):
            mnist_instance = MnistInstance(image=PIL.Image.open(os.path.join(directory_path, img_path)),
                                           label=int(os.path.splitext(img_path)[0].split('_')[1]))
            bag.append(mnist_instance)

        return bag

    @property
    def images(self):
        return [mnist.image for mnist in self]

    def digit_in_bag(self, digit: int):
        return any(mnist.label == digit for mnist in self)

    def append(self, __object: MnistInstance) -> None:
        if not isinstance(__object, MnistInstance):
            raise Exception(type(__object))
        super(Bag, self).append(__object)

    def bag_dir(self, save_path: str) -> str:
        bag_name = "_".join([str(item.label) for item in self])
        if self.label is not None:
            bag_dir = os.path.join(save_path, str(self.label), bag_name)
        else:
            bag_dir = os.path.join(save_path, bag_name)

        return bag_dir

    def save(self, save_path: str):
        bag_dir = self.bag_dir(save_path=save_path)
        os.makedirs(bag_dir, exist_ok=True)
        for position, item in enumerate(self):
            item: MnistInstance
            _item_count = 0
            item_name = f"{position}_{item.label}_{_item_count}.jpg"
            while os.path.exists(os.path.join(bag_dir, item_name)):
                _item_count += 1
                item_name = f"{position}_{item.label}_{_item_count}.jpg"
            item.image.save(os.path.join(bag_dir, item_name))


class BagCollections(list):
    def __init__(self, *items):
        super().__init__()
        for __item in items:
            self.append(__item)

    def append(self, __object: Bag) -> None:
        if not isinstance(__object, Bag):
            raise Exception()
        super(BagCollections, self).append(__object)


class MnistBagsDataset(data.Dataset):
    DECISION_DIGIT = 7

    def __init__(self, root: str, number_of_bags: int, train: bool = True,
                 transforms: Optional[torchvision.transforms.Compose] = None):
        self.mnist_data = torchvision.datasets.MNIST(root=root, train=train, download=True)
        self.number_of_bags = number_of_bags
        self.transforms = transforms

        self.mnist_data_all_ids = self._prepare_mnist_data_all_ids()
        self.mnist_data_without_decision_digit_ids = self._prepare_mnist_data_without_decision_digit_ids()

        self.__bags = BagCollections()

    @property
    def bags(self):
        return self.__bags

    def __len__(self):
        return len(self.bags)

    def _prepare_mnist_data_all_ids(self) -> list[int]:
        return list(range(len(self.mnist_data.targets)))

    def _prepare_mnist_data_without_decision_digit_ids(self) -> list[int]:
        return [idx for idx in range(len(self.mnist_data.targets)) if
                self.mnist_data.targets[idx] != MnistBagsDataset.DECISION_DIGIT]

    def _prepare_positive_bag(self) -> Bag:
        bag_size = BagDistribution.get_bag_size()
        bag_mnist_ids = random.sample(self.mnist_data_all_ids, bag_size)
        bag = Bag([MnistInstance(*self.mnist_data[idx]) for idx in bag_mnist_ids])
        while not bag.digit_in_bag(MnistBagsDataset.DECISION_DIGIT):
            bag_mnist_ids = random.sample(self.mnist_data_all_ids, bag_size)
            bag = Bag([MnistInstance(*self.mnist_data[idx]) for idx in bag_mnist_ids])
        bag.label = 1

        return bag

    def _prepare_positive_bags(self):
        number_of_positive_bags = int(self.number_of_bags / 2)
        for _ in range(number_of_positive_bags):
            self.__bags.append(self._prepare_positive_bag())

    def _prepare_negative_bag(self) -> Bag:
        bag_size = BagDistribution.get_bag_size()
        bag_mnist_ids = random.sample(self.mnist_data_without_decision_digit_ids, bag_size)
        bag = Bag([MnistInstance(*self.mnist_data[idx]) for idx in bag_mnist_ids])
        bag.label = 0

        return bag

    def _prepare_negative_bags(self):
        number_of_negative_bags = int(self.number_of_bags / 2)
        for _ in range(number_of_negative_bags):
            self.__bags.append(self._prepare_negative_bag())

    def _remove_bags(self):
        self.__bags = BagCollections()

    def _prepare_bags(self):
        self._prepare_positive_bags()
        self._prepare_negative_bags()

    def regenerate_bags(self):
        self._remove_bags()
        self._prepare_bags()

    def save_bags(self, save_path: str):
        for bag in self.bags:
            bag.save(save_path=save_path)

    def __getitem__(self, item: int):
        bag = self.bags[item]
        bag_images = bag.images

        if self.transforms:
            bag_images = [self.transforms(img) for img in bag_images]
        bag_images = torch.stack(bag_images)
        label = torch.from_numpy(np.array([bag.label])).to(torch.int64)

        return bag_images, label


def prepare_train_dataloader(number_of_bags: int = 1000, batch_size: int = 16):
    train_transforms = torchvision.transforms.Compose([
        torchvision.transforms.ToTensor(),
    ])
    train_data = MnistBagsDataset(root=os.path.join('data'),
                                  train=True,
                                  number_of_bags=number_of_bags,
                                  transforms=train_transforms)
    train_data.regenerate_bags()
    loader = data.DataLoader(dataset=train_data,
                             batch_size=batch_size,
                             shuffle=True,
                             collate_fn=collate_fn
                             )

    return loader


def collate_fn(_data):
    inputs = [d[0] for d in _data]
    labels = [d[1] for d in _data]
    inputs = pad_sequence(inputs, batch_first=True)
    labels = torch.tensor(labels)

    return inputs, labels


def prepare_val_dataloader(number_of_bags: int = 100, batch_size: int = 16):
    val_transforms = torchvision.transforms.Compose([
        torchvision.transforms.ToTensor(),
    ])
    val_data = MnistBagsDataset(root=os.path.join('data'),
                                train=True,
                                number_of_bags=number_of_bags,
                                transforms=val_transforms)
    val_data.regenerate_bags()
    loader = data.DataLoader(dataset=val_data,
                             batch_size=batch_size,
                             shuffle=False,
                             collate_fn=collate_fn)

    return loader


weakly_supervised_classification/model/model_evaluator.py
import glob
import os

import PIL.Image
import torch
import torchvision

from weakly_supervised_classification.model.data_loader import Bag


class Evaluator:
    def __init__(self, model_path: str, batch_size: int = 64):
        self.batch_size = batch_size
        self.model = self.load_model(model_path=model_path)

        self.transforms = torchvision.transforms.Compose([
            torchvision.transforms.ToTensor(),
        ])

    @staticmethod
    def load_model(model_path: str):
        """
        Loads model model from specified path.
        :param model_path: (str) path to save TorchScript model
        :return: model object
        """
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

        return torch.jit.load(model_path, map_location=device)

    def _preprocess_batch(self, images_batch):
        bag_images = [self.transforms(img) for img in images_batch]
        bag_images = torch.stack(bag_images)
        bag_images = bag_images.unsqueeze(0)

        return bag_images

    def evaluate_dir(self, images_dir):
        bag = Bag.from_directory(images_dir)

        return self.evaluate_bag(bag)

    def evaluate_bag(self, bag: Bag):
        batch = bag.images

        return self.evaluate_batch(batch)

    @torch.no_grad()
    def evaluate_batch(self, images_batch):
        self.model.eval()

        images_batch = self._preprocess_batch(images_batch)
        outputs = []
        for batch in images_batch.split(self.batch_size):
            outputs += self.model(batch.to(next(self.model.parameters()).device))

        return outputs


weakly_supervised_classification/model/multi_attention_mil.py
import torch
import torch.nn as nn
import torchvision.models


class Normalize(torch.nn.Module):
    """
    Normalization module for debiasing and normalizing input data values in range [0, 1] with commonly used
    normalization values.
    """
    def __init__(self):
        super(Normalize, self).__init__()
        self.mean = torch.nn.Parameter(torch.as_tensor([0.1307])[None, :, None, None], requires_grad=False)
        self.std = torch.nn.Parameter(torch.as_tensor([0.3081])[None, :, None, None], requires_grad=False)

    def forward(self, x):
        x = x - self.mean
        x = x / self.std
        return x


class Attention(nn.Module):
    """
    Attention module
    """
    def __init__(self, _dim: int = 128):
        super().__init__()
        self._D = _dim
        self.attention = nn.Sequential(
            nn.Linear(self._D, self._D), nn.Tanh(), nn.Linear(self._D, 1)
        )

        self.fc = nn.Sequential(
            nn.Linear(self._D, self._D),
            nn.ReLU(),
        )

    def forward(self, _input):
        feature = self.fc(_input)

        attention = self.attention(feature)
        attention = nn.Softmax(dim=1)(attention)
        attention = torch.transpose(attention, 2, 1)

        m = torch.matmul(attention, feature)
        m = m.view(-1, 1 * self._D)

        return feature, attention, m


class FirstAttention(Attention):
    """
    First attention module
    """
    def __init__(self, fc_layer_size: int = 512, _dim: int = 128):
        super().__init__(_dim=_dim)
        self.fc_layer_size = fc_layer_size

        self.fc = nn.Sequential(
            nn.Linear(self.fc_layer_size, self._D),
            nn.ReLU(),
        )


class MultiAttentionMIL(nn.Module):
    """
    Multi Attention Multiple Instance Learning module
    """
    def __init__(self, num_classes: int = 2, num_attentions: int = 3, fc_layer_size: int = 512):
        super().__init__()
        self.num_classes = num_classes
        self.num_attentions = num_attentions

        self.__num_features = 512
        self.fc_layer_size = fc_layer_size
        self._D = 128

        self.model = torchvision.models.resnet18()
        self.model.conv1 = nn.Conv2d(1, 64, kernel_size=7, stride=2, padding=3, bias=False)
        self.model.fc = torch.nn.Sequential(torch.nn.Linear(self.__num_features, self.fc_layer_size))

        self.normalization = Normalize()
        self.first_attention = FirstAttention(fc_layer_size=fc_layer_size, _dim=self._D)
        self.attention = Attention(_dim=self._D)

        self.fc_result = nn.Sequential(nn.Linear(self._D, self.num_classes),
                                       nn.Sigmoid())

    def forward(self, images):
        images = self.normalization(images)
        feature = torch.stack([self.model(__batch) for __batch in images])
        feature = feature.squeeze(-1).squeeze(-1)

        feature, attention, m = self.first_attention(feature)
        m_tot = m

        for num in range(self.num_attentions):
            feature, attention, m = self.attention(feature)
            m_tot += m
        result = self.fc_result(m_tot)

        return result, attention


weakly_supervised_classification/main.py
import os
import random
import argparse

from weakly_supervised_classification.model.data_loader import MnistBagsDataset
from weakly_supervised_classification.model.tester import Tester
from weakly_supervised_classification.model.trainer import Trainer


def prepare_test_sets():
    random.seed(12345)
    parser = argparse.ArgumentParser('Preparing two test datasets for weakly-supervised-learning',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--bags', help='Amount bags for each dataset', type=int, default=10000)
    parser.add_argument('--output-directory', help='Output directory path', type=str, default='data')
    args = parser.parse_args()

    test_path_1 = os.path.join(args.output_directory, "test_1")
    if os.path.exists(test_path_1):
        raise Exception(f"Test dataset 1 exists in {test_path_1}")
    os.makedirs(test_path_1)

    test_path_2 = os.path.join(args.output_directory, "test_2")
    if os.path.exists(test_path_2):
        raise Exception(f"Test dataset 2 exists in {test_path_2}")
    os.makedirs(test_path_2)

    test_data_1 = MnistBagsDataset(root=os.path.join('data'),
                                   train=False,
                                   number_of_bags=args.bags)
    test_data_1.regenerate_bags()
    test_data_1.save_bags(test_path_1)

    test_data_2 = MnistBagsDataset(root=os.path.join('data'),
                                   train=False,
                                   number_of_bags=args.bags)
    test_data_2.regenerate_bags()
    test_data_2.save_bags(test_path_2)

    print(f'Save {len(test_data_1)} test bags to {test_path_1} and '
          f'{len(test_data_2)} test bags to {test_path_2}')


def train_test_models():
    parser = argparse.ArgumentParser('Learning and testing models for weakly-supervised-learning',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--training-bags', help='Amount bags for each training epoch', type=int, default=200)
    parser.add_argument('--validation-bags', help='Amount bags for validation', type=int, default=1000)
    parser.add_argument('--attentions', help='Amount additional attentions in model', type=int, default=2)
    parser.add_argument('--epochs', help='Training epochs', type=int, default=200)

    parser.add_argument('--models', help='Models amount', type=int, default=10)
    parser.add_argument('--lr', help='Learning rate', type=float, default=0.00005)
    parser.add_argument('--output-directory', help='Output directory path', type=str,
                        default='checkpoints')
    parser.add_argument('--test-data', help='Testing data path', type=str,
                        default='data/test_1')

    args = parser.parse_args()

    for num_model in range(args.models):
        trainer = Trainer(number_train_of_bags=args.training_bags, number_eval_of_bags=args.validation_bags,
                          num_attentions=args.attentions, lr=args.lr,
                          models_path=os.path.join(args.output_directory, str(num_model)))
        trainer.train(args.epochs)
    tester = Tester(models_path=args.output_directory, data_path=args.test_data, output_path='results')
    tester.test()


def test_models():
    parser = argparse.ArgumentParser('Testing models for weakly-supervised-learning',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--models-directory', help='Models directory, or path to model directory.',
                        type=str, default='checkpoints')
    parser.add_argument('--save-correct-attentions', help='Saves all attentions images', action='store_true')
    parser.add_argument('--output-directory', help='Output directory path', type=str,
                        default='results_best_model')
    parser.add_argument('--test-data', help='Testing data path', type=str,
                        default='data/test_2')

    args = parser.parse_args()
    tester = Tester(models_path=args.models_directory, data_path=args.test_data, output_path=args.output_directory,
                    save_correct_attentions=args.save_correct_attentions)
    tester.test()
